### PR TITLE
WIP InteractiveViewer Widget

### PR DESCRIPTION
**Pull request by justinmc**
## Description

This PR introduces a new widget called InteractiveViewer.  Any widget passed as its child will be able to be panned, zoomed, and rotated by user gestures.

### Image viewer example
This example can be seen in the Flutter Gallery under Material -> Grid.

```dart
Scaffold(
  body: SizedBox.expand(
    child: Hero(
      tag: photo.tag,
      child: InteractiveViewer(
        disableRotation: true,
        child: Image.asset(
          photo.assetName,
          package: photo.assetPackage,
          fit: BoxFit.cover,
        ),
      ),
    ),
  ),
);
```

### Game board example
This example is also in the Flutter Gallery, under Studies -> 2D Transformations.

```dart
Scaffold(
  body: LayoutBuilder(
    builder: (BuildContext context, BoxConstraints constraints) {
      // Draw the scene as big as is available, but allow the user to
      // translate beyond that to a visibleSize that's a bit bigger.
      final Size size = Size(constraints.maxWidth, constraints.maxHeight);
      final Size visibleSize = Size(size.width * 3, size.height * 2);
      return InteractiveViewer(
        reset: _reset,
        onResetEnd: () {
          setState(() {
            _reset = false;
          });
        },
        child: CustomPaint(
          painter: painter,
        ),
        boundaryRect: Rect.fromLTWH(
          -visibleSize.width / 2,
          -visibleSize.height / 2,
          visibleSize.width,
          visibleSize.height,
        ),
        // Center the board in the middle of the screen. It's drawn centered
        // at the origin, which is the top left corner of the
        // InteractiveViewer.
        initialTranslation: Offset(size.width / 2, size.height / 2),
        onTapUp: _onTapUp,
      );
    },
  ),
),
```

## Tests

TODO

## Other TODOs
  * A boolean flag that locks an individual scroll event to one direction only (as in a spreadsheet).
  * Update flutter_gallery repo to use InteractiveViewer (see gesture-transformable branch in that repo).

### CODE REVIEW

**Comment by pilt** [comment]
Thanks for this!

I'm playing around with it and have DragTarget widgets as descendants to InteractiveViewer. It seems like the only drag targets that are willing to accept draggables are those that were visible on first render. Draggables that weren't visible by then but are now because of panning or zooming out won't accept draggables.

The top-level child of my InteractiveViewer is an OverflowBox. I'm quite new to Flutter and don't know if the problem could be caused by this, as I don't know of another way to render standard/flex widgets within the interactive viewer without getting overflow errors (I render a grid using Column and Row widgets inside the overflow box with manually specified height and width).

I've attempted to force a re-render of draggables by mutating their state when the drag action starts but it didn't make a difference.

Any idea what the cause might be? Would it help if I created a minimal example?

Thanks again.

**Comment by pilt** [comment]
I found a workaround. I need only pan and zoom for my use case, and this makes all drag targets work. It might be ugly but I'm unable to tell. 

First, I got the idea from the discussion in https://github.com/flutter/flutter/issues/27587.

I use Positioned and Transform.scale to render the child of InteractiveViewer (and I could get rid of the OverflowBox and just use a Container with width and height, but I don't know if that matters):

```dart
      child: Stack(
        children: <Widget>[
          Positioned(
            left: widget.transformationController.value.row0.w,
            top: widget.transformationController.value.row1.w,
            child: KeyedSubtree(
              key: _childKey,
              child: Transform.scale(
                alignment: Alignment.topLeft,
                scale: widget.transformationController.value.row0.x,
                child: widget.child,
              ),
            ),
          ),
        ],
      ),
```

**Comment by justinmc** [comment]
@pilt Thanks for the input on this, I'm sure other people will try to put drop targets in this.  In your final solution are you modifying the code in this PR, or did you build something totally separate?

File: packages/flutter/lib/src/widgets/interactive_viewer.dart [hunk 1]
```diff
@@ -0,0 +1,1085 @@
...
+      // Wrapping a Widget in an InteractiveViewer does not change how the
+      // widget is initially rendered. It should look identical whether or not
+      // the InteractiveViewer is there, until the transformation is changed.
+      child: Transform(
+        transform: widget.transformationController.value,
+        child: KeyedSubtree(
+          key: _childKey,
+          child: widget.child,
+        ),
+      ),
```
Comment (pilt) [inline packages/flutter/lib/src/widgets/interactive_viewer.dart]: Here's where I'm using this snippet instead (I don't need rotation):

````dart
      child: Stack(
        children: <Widget>[
          Positioned(
            left: widget.transformationController.value.row0.w,
            top: widget.transformationController.value.row1.w,
            child: KeyedSubtree(
              key: _childKey,
              child: Transform.scale(
                alignment: Alignment.topLeft,
                scale: widget.transformationController.value.row0.x,
                child: widget.child,
              ),
            ),
          ),
        ],
      ),
```

Comment (justinmc) [reply packages/flutter/lib/src/widgets/interactive_viewer.dart]: Ah I see what you mean, thanks.  It seems like translation and rotation with the Transform widget cause drop targets to not work.  I'll try to look into this before I release this code.

**Comment by justinmc** [comment]
This has gotten out of date, so I'll need to re-create the PR.  I'm closing it, and will link to the new PR when I open it.

**Comment by TzviPM** [comment]
I'm guessing this was the follow-up PR? https://github.com/flutter/flutter/pull/56409

**Pull request closed without merging**